src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -0,0 +1,248 @@
+                                break;
+                            }
+                        }
+                    }
+
+                    // Check if executing actions has been throttled due to count limit
+                    if (actionCountLimit > executionTimestamps.Count)
+                    {
+                        executionTimestamps.Enqueue(currentTimestamp);
+
+                        try
+                        {
+                            // Intentionally not using the linkedToken. Allow the action list to execute gracefully
+                            // unless forced by a caller to cancel or stop the running of the pipeline.
+                            await _actionListExecutor.ExecuteActions(_ruleOptions.Actions, _endpointInfo, token);
+                        }
+                        catch (Exception ex) when (ex is not OperationCanceledException)
+                        {
+                            // Bad action execution shouldn't fail the pipeline.
+                            // TODO: log that the action execution has failed.
+                        }
+                        finally
+                        {
+                            // The collection rule has executed the action list the maximum
+                            // number of times as specified by the limits and the action count
+                            // window was not specified. Since the pipeline can no longer execute
+                            // actions, the pipeline can complete.
+                            completePipeline = actionCountLimit <= executionTimestamps.Count &&
+                                !actionCountWindowDuration.HasValue;
+                        }
+                    }
+                    else
+                    {
+                        // Throttled
+                    }
+
+                    linkedToken.ThrowIfCancellationRequested();
+
+                    // If the trigger is a startup trigger, only execute the action list once
+                    // and then complete the pipeline.
+                    if (trigger is ICollectionRuleStartupTrigger)
+                    {
+                        // Signal that the pipeline trigger is initialized.
+                        _startedSource.TrySetResult(null);
+
+                        // Complete the pipeline since the action list is only executed once
+                        // for collection rules with startup triggers.
+                        completePipeline = true;
+                    }
+                }
+            }
+            catch (OperationCanceledException) when (durationCancellationSource.IsCancellationRequested)
+            {
+                // This exception is caused by the pipeline duration expiring.
+                // Handle it to allow pipeline to be in completed state.
+            }
+        }
+
+        protected override Task OnCleanup()
+        {
+            _startedSource.TrySetCanceled();
```

Comment:
iirc OnCleanup will always happen not necessarily due to a cancellation.

Reply:
That's fine because this is using TrySetCanceled. If a caller creates the pipeline, calls StartAsync, and disposes the pipeline before StartAsync completes, then the pipeline needs to cancel the started source so that the StartAsync method is canceled.